File: python-algorithm/ds_1_stack_queue/L84-h.py
# -*- coding: utf-8 -*-


# 单调递增栈，即出现破坏单调性的元素的时候，找它的上一个元素往起始方向
# 滑动可组成的最大矩形，直到高度≤破坏单调性的元素的高度（即栈恢复单调递增）

class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        # 为了让最两端的元素有可比较值，让它们出栈，两端加2个0
        heights.insert(0, 0)
        heights.append(0)
        stk = []
        ans = 0
        for i in range(len(heights)):
            while stk != [] and heights[stk[-1]]>heights[i]:
                cur_idx = stk.pop(-1)
                # 容易出错，左边界left不能直接使用cur_idx，因为往左可能隔着破坏单调性的元素
                left = stk[-1] + 1
                right_border = i-1
                ans = max(ans, (right_border-left+1)*heights[cur_idx])
            stk.append(i)
        return ans

File: ds_1_stack_queue/L84-h.cpp


// 这道题设计的精巧，比较难；需要在原来的高度数组两两边各添加一个0，用来确定每个出栈元素的左右边界范围
// 每次出栈计算一次面积；每次栈的单调递增性被破坏的时候开始出栈，一直出栈到单调性恢复为止.
// 这个题不太好理解，可以一步步看下过程。


#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    int largestRectangleArea(vector<int>& heights){
        int ans = 0;
        vector<int> st;
        heights.insert(heights.begin(), 0);
        heights.push_back(0);
        for (int i = 0; i < heights.size(); i++)
        {
            while (!st.empty() && heights[st.back()] > heights[i])
            {
                int cur = st.back();
                st.pop_back();
                int left = st.back()+1 ; //左边界的计算一定要这样
                int right = i - 1;
                ans = max(ans, (right - left + 1) * heights[cur]);
            }
            st.push_back(i);
        }
        return ans;
    }

};

#include <stack>
class Solution1 {
public:
    int largestRectangleArea(vector<int>& heights){
        int ans = 0;
        stack<int> stk;
        heights.insert(heights.begin(), 0);
        heights.push_back(0);
        for (int i = 0; i < heights.size(); i++)
        {
            while (!stk.empty() && heights[stk.top()] > heights[i])
            {
                int cur = stk.top();
                stk.pop();
                int left = stk.top()+1 ; //左边界的计算一定要这样
                int right = i - 1;
                ans = max(ans, (right - left + 1) * heights[cur]);
            }
            stk.push(i);
        }
        return ans;
    }

};

int main(){
    vector<int> vec{2,1,5,6,2,3};
    cout << Solution().largestRectangleArea(vec)<<endl;
    cout << Solution1().largestRectangleArea(vec)<<endl;
}